astropy/units/typing_utils.py

```diff
@@ -1,116 +1,64 @@
-        obj = getattr(module, name)
-        # find all unit-like objects
-        if isinstance(obj, (UnitBase, FunctionUnitBase)):
-            cls = type(obj)
-            type_name = cls.__name__
-            docstring = f'"""{obj.__doc__}"""\n' if has_own_docstring(obj) else ""
-            variable_lines.append(f"{name}: {type_name}\n{docstring}")
-            # the class of the unit object (e.g., IrreducibleUnit) is often
-            # not defined in the module itself, so we need to import it
-            import_lines.add(make_import_line(cls))
-
-    # --- output ---
-    yield "# This stub file was automatically generated."
-    yield ""
-    yield from sorted(import_lines)
-    yield ""
-    yield from variable_lines
+__all__: list[str] = []
 
-
-def make_import_line(cls) -> str:
-    """Create a "from ... import ..." line for a given class."""
-    return f"from {cls.__module__} import {cls.__name__}"
-
-
-def has_own_docstring(obj) -> bool:
-    """Check if object has its own docstring vs an inherited one."""
-    if not hasattr(obj, "__doc__") or not obj.__doc__:
-        return False
-
-    # for classes, check if docstring differs from parent classes
-    if isinstance(obj, type):
-        for base in obj.__bases__:
-            if hasattr(base, "__doc__") and base.__doc__ == obj.__doc__:
-                return False
-
-    # for instances, check if docstring differs from class
-    elif hasattr(obj, "__class__"):
-        if hasattr(obj.__class__, "__doc__") and obj.__class__.__doc__ == obj.__doc__:
-            return False
-
-    return True
-
-
-def nfkc_normalize(items: Iterable[str]) -> Iterator[str]:
-    yield from (unicodedata.normalize("NFKC", item) for item in items)
-
-
-def unique(items: Iterable) -> Iterator:
-    """Iterate over unique items while preserving order."""
-    seen = set()
-    for item in items:
-        if item in seen:
-            continue
-        seen.add(item)
-        yield item
+import inspect
+from pathlib import Path
+from types import ModuleType
+
+from astropy import units as u
+from astropy.units import cds
+
+
+def generate_stub(module: ModuleType, output_dir: Path) -> None:
+    members = {name: type(getattr(module, name)) for name in module.__all__}
+    functions = {}
+    for name, cls in tuple(members.items()):
+        if cls.__name__ == "function":
+            members.pop(name)
+            functions[name] = inspect.get_annotations(getattr(module, name))
+    import_lines = sorted(
+        f"from {cls.__module__} import {cls.__name__}\n"
+        for cls in set(members.values()) | {ann["return"] for ann in functions.values()}
+    )
+    with (
+        output_dir.joinpath(*module.__name__.split("."))
+        .with_suffix(".pyi")
+        .open("wt", encoding="utf-8") as stub_file
+    ):
+        stub_file.write(
+            "# This stub file was automatically generated.\n\n__all__ = [\n"
+        )
+        stub_file.writelines(f'    "{name}",\n' for name in module.__all__)
```

Comment:
What is the point of copying `__all__` to the stub file? All entries in it are guaranteed to be in `module.__all__` anyway.

Reply:
In that case, little choice I guess. Just seems a bad design, as it doubles the number of lines in the file for no reason.

Reply:
An explicit `__all__` allows distinguishing what the module in question intends to expose from what was imported just to be used in annotations.

Reply:
Ah, yes, not thinking clearly enough. Explicit `__all__` is OK then.